service/src/main/java/org/cbioportal/genome_nexus/service/enricher/IsoformAnnotationEnricher.java

```diff
@@ -1,95 +1,84 @@
-import org.cbioportal.genome_nexus.service.exception.IsoformOverrideNotFoundException;
+import org.cbioportal.genome_nexus.service.EnsemblService;
+import org.cbioportal.genome_nexus.util.IsoformOverrideSource;
 
-import java.util.LinkedList;
+import java.util.Collections;
 import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
 
-/**
- * @author Selcuk Onur Sumer
- */
 public class IsoformAnnotationEnricher extends BaseAnnotationEnricher
 {
     String source;
-    IsoformOverrideService service;
+    EnsemblService ensemblService;
 
     public IsoformAnnotationEnricher(
         String id,
         String source,
-        IsoformOverrideService service
+        EnsemblService ensemblService
     ) {
         super(id);
         this.source = source;
-        this.service = service;
+        this.ensemblService = ensemblService;
     }
 
     @Override
     public void enrich(VariantAnnotation annotation)
     {
         // no transcripts to enrich, abort.
-        if (annotation.getTranscriptConsequences() == null)
-        {
+        if (annotation.getTranscriptConsequences() == null) {
             return;
         }
 
-        List<TranscriptConsequence> transcripts = new LinkedList<>();
+        // get transcript overrides for the given isoform override source
+        // if no isoform override source is provided, then the default is uniprot
+        Set<String> predefinedCanonicalTranscriptIds = this.ensemblService.getCanonicalTranscriptIdsBySource(
+            IsoformOverrideSource.getOrDefault(this.source)
+        );
 
-        /* This logic handles the case when there are more than one potential genes that a variant could be affecting.
-            When two genes are extremely close in the genome/overlapping, VEP will return the transcripts for both genes.
-            In the case where we only have an override for one of the genes, we do not want to force those transcripts as canonical
-            if the most severe consequence does not exist in that gene, because more likely the correct transcript that the user
-            cares about exists in the other gene.
+        List<TranscriptConsequence> canonicalTranscriptCandidates = this.findCanonicalTranscriptCandidates(
+            annotation.getTranscriptConsequences(),
+            predefinedCanonicalTranscriptIds
+        );
 
-            We only apply isoform overrides if the most severe consequence exists in the gene we want to apply to override to.
-            If it does not, do not touch the canonical fields for those transcripts.
-        */
-        String geneWithMostSevereConsequence = "";
-        String mostSevereConsequence = annotation.getMostSevereConsequence();
-        for (TranscriptConsequence transcript : annotation.getTranscriptConsequences()) {
-            if (transcript.getConsequenceTerms().contains(mostSevereConsequence)) {
-                geneWithMostSevereConsequence = transcript.getGeneSymbol();
-                break;
```

Comment:
@inodb we don't need this logic anymore, right?

Reply:
@onursumer yeah i think we can remove it but note that the logic used to be different. I.e. to me it seems like (1) it looks for the most severe consequence and then if that is in a gene for which we have an override then we apply it. Which is different from what we are doing now: (2) filter to canonical transcripts and then get most severe event in that subset. I think for our case of annotating for the purpose of loading it into cBioPortal (2) makes more sense, b/c if it is not in a canonical transcript it means there is no associated hugo symbol and no one can even query that mutation in cBioPortal.

Reply:
I think we touched on this in our discussion yesterday. The case was two overlapping genes on opposite strands .. and also a mutation in an intergenic region near these genes. I think it would be a good idea (in the near future) to consider and formalize (in a util / service) our logic for how to select the gene we "want" to base our annotation on when there is ambiguity. Queries to the annotation service only provide chromosome coordinates, but potentially we could also accept a [optional?] gene choice which would aid in the disambiguation. But even if we do not want to do this, it would be good to have a coded (and unit tested) component which takes the ambiguous (multi-gene) transcripts coming out of VEP and applies our "pick a gene" logic to it. Once we have chosen the intended gene, we can filter the TranscriptConsequences to the set of consequences related to that gene's transcripts and then proceed to select the transcript. This way the same gene selection logic would be applied regardless of whether the canonical transcript is identified by the table in Mongo or is identified by the "canonical" element in the VEP response.
For now, when there are multiple genes in the VEP response we are changing behavior so that we prefer genes from the override set (in mongo). However, there are probably cases where this is not enough to fully disambiguate.
Also, I wonder whether every hugo symbol from the gene table has been added to the Mongo table? The final statement from @inodb above seems to imply that. If we want to rely on that, we should be careful to keep the Mongo table in sync with the gene table as there are gene table updates.